### PR TITLE
Use system fonts as fallback.

### PR DESCRIPTION
**Pull request by bruvzg**
- Add support for font weight and stretch selection when using system fonts.
- Add OS function to get system fallback font from a font name, text, and language code (`OS.get_system_font_path_for_text`). 
- Automatically use system fonts as fallback (as a last resort, after all user specified fallback fonts fail), system font is selected by OS, based on the main font name and style.

Implemented on Android, iOS, Linux (fontconfig required), macOS, and Windows.

For the reference: Implementing it for Web is only possible with https://wicg.github.io/local-font-access/ (draft proposal, not supported by anything).

Test Project:
[FontSelect.zip](https://github.com/godotengine/godot/files/10098482/FontSelect.zip)

### CODE REVIEW

**Comment by bruvzg** [comment]
Performance seems to be pretty reasonable, font lookup takes about 100-1500 μs (every time text with the missing glyphs is shaped, Linux version is the slowest). Loading all fonts for the sample on the screenshot (6 to 10 fonts depending on OS) takes about 15000 μs on average. So should be fine for UI apps (or as fallback anywhere).

Tested with the empty project, a `Label`, and no fonts configured:
<details>
  <summary>Screenshots</summary>

|          |  |
|--------------|:-----:|
| Linux (Pop!_OS 22.04) | linux |
| macOS 13.0.1 | macos |
| Windows 11 | windows |

*It's not a clean OS install, so the default font set might be different.*

</details>

I'll take a look if it's possible to do it on Android (I do not think there's an official API, but should be possible to read system configs directly) and it should be good to go.

File: doc/classes/FileAccess.xml [hunk 1]
```diff
@@ -153,6 +153,20 @@
 				Returns the last error that happened when trying to perform operations. Compare with the [code]ERR_FILE_*[/code] constants from [enum Error].
 			</description>
 		</method>
+		<method name="get_file_as_array" qualifiers="static">
+			<return type="PackedByteArray" />
+			<param index="0" name="path" type="String" />
+			<description>
+				Returns the whole [param path] file contents as a [PackedByteArray]
```
Comment (akien-mga) [inline doc/classes/FileAccess.xml:160]:
```suggestion
				Returns the whole [param path] file contents as a [PackedByteArray].
```

Comment (akien-mga) [reply doc/classes/FileAccess.xml:160]: Maybe should also clarify the encoding?

File: platform/android/java/lib/src/org/godotengine/godot/io/StorageScope.kt [hunk 2]
```diff
@@ -90,6 +90,15 @@ internal enum class StorageScope {
...
 			}
 
+			var rootDir: String? = System.getenv("ANDROID_ROOT")
+			if (rootDir != null && canonicalPathFile.startsWith(rootDir)) {
+				return APP
+			}
+
+			if (canonicalPathFile.startsWith("/system/")) {
+				return APP
+			}
```
Comment (akien-mga) [inline platform/android/java/lib/src/org/godotengine/godot/io/StorageScope.kt:100]: CC @m4gr3d

File: servers/text_server.h [hunk 3]
```diff
@@ -498,6 +507,8 @@ class TextServer : public RefCounted {
 
 	TypedArray<Vector2i> parse_structured_text(StructuredTextParser p_parser_type, const Array &p_args, const String &p_text) const;
 
+	virtual void cleanup(){};
```
Comment (akien-mga) [inline servers/text_server.h:510]:
```suggestion
	virtual void cleanup() {}
```

File: doc/classes/FileAccess.xml (hunk 1 above)
Comment (bruvzg) [reply doc/classes/FileAccess.xml:160]: Encoding? It's reading the file as is (it's for primary for reading binary files, not text).

Comment (akien-mga) [reply doc/classes/FileAccess.xml:160]: Ah right, makes sense.

Comment (bruvzg) [reply doc/classes/FileAccess.xml]: Added `without any decoding.` to clarify it's reading raw bytes.

File: core/io/file_access.cpp [hunk 4]
```diff
@@ -810,6 +810,9 @@ void FileAccess::_bind_methods() {
 	ClassDB::bind_static_method("FileAccess", D_METHOD("open_compressed", "path", "mode_flags", "compression_mode"), &FileAccess::open_compressed, DEFVAL(0));
 	ClassDB::bind_static_method("FileAccess", D_METHOD("get_open_error"), &FileAccess::get_open_error);
 
+	ClassDB::bind_static_method("FileAccess", D_METHOD("get_file_as_array", "path"), &FileAccess::_get_file_as_array);
```
Comment (KoBeWi) [inline core/io/file_access.cpp:813]: I think `get_file_as_bytes()` would be a better name.

Comment (akien-mga) [reply core/io/file_access.cpp:813]: Sounds good to me.

**Review by akien-mga — APPROVED** [review]
Looks good to me.

There's a suggestion from KoBeWi for the new File API, to be discussed. Otherwise the rest should be good to merge.

File: platform/android/java/lib/src/org/godotengine/godot/io/StorageScope.kt (hunk 2 above)
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/io/StorageScope.kt:100]: Aren't `/system` and `System.getenv("ANDROID_ROOT")` the same value? Do we need this second check?

**Review by m4gr3d — CHANGES_REQUESTED** [review]

File: core/io/file_access.cpp (hunk 4 above)
Comment (bruvzg) [reply core/io/file_access.cpp]: Renamed (it's not a new API, but was not exposed, so should not be a breaking change).

File: platform/android/java/lib/src/org/godotengine/godot/io/StorageScope.kt (hunk 2 above)
Comment (bruvzg) [reply platform/android/java/lib/src/org/godotengine/godot/io/StorageScope.kt]: Seems like it is the same, removed extra `/system` check.

**Review by m4gr3d — APPROVED** [review]
The Android section looks good to me.

**Comment by akien-mga** [comment]
Thanks!

**Comment by fire** [comment]
I believe this changes "fontconfig" from optional to required on Linux. I'm not debugging my system today.

**Comment by alexzheng** [comment]
Nice feature.
Will it back port to 3.x?

**Comment by Calinou** [comment]
> Nice feature. Will it back port to 3.x?

System font support relies on a lot of backwards-incompatible changes, so it's not possible to backport this feature without a complete rewrite.

**Pull request merged**